lib/ansible/galaxy/collection/__init__.py

```diff
@@ -769,6 +769,19 @@ def install_collections(
                     "Skipping signature verification."
                 )
 
+            if concrete_coll_pin.type == 'galaxy':
```

Comment:
Is this only useful during installation? It seems, such a conversion might belong to `_resolve_depenency_map()` or even deeper, somewher here: https://github.com/ansible/ansible/blob/9faf944/lib/ansible/galaxy/dependency_resolution/providers.py#L376.

If the signatures are taken into consideration by the depresolver itself, wouldn't it cut the undesired backtracking path early?

Reply:
Yes, just installation. `_resolve_depenency_map` is used by download too and it's not needed there.

Maybe I misunderstand  what you mean, but signatures are not taken into account by the dependency resolver (besides compiling extra signatures which are user-provided and don't add an API call) and would basically prevent this bugfix (unless we do something like overriding `CollectionDependencyResolver.resolve()` anyway to handle it after find_matches).

We could consider signature validity + versions in the dependency resolver (a feature?), but backtracking if signature validation fails seems like it could be problematic (a change in behavior at least) - if someone prefers to install collections from AH but fall back to Galaxy for collections missing from AH; then any collections on AH that fail signature validation may not give an indicator something was wrong if an unsigned version of the collection was found on Galaxy.

Reply:
> Yes, just installation. `_resolve_depenency_map` is used by download too and it's not needed there.

I'd argue that when, verification is enabled, download shouldn't place files on disk either. Hence, I'd expect it to be needed there... Perhaps, this is out of the scope of this specific PR, but in general, it seems like a good idea.

> Maybe I misunderstand what you mean, but signatures are not taken into account by the dependency resolver (besides compiling signatures - all of them on devel, and now on this branch just those which are user-provided and don't add an API call) and would basically prevent this bugfix (unless we do something like overriding `CollectionDependencyResolver.resolve()` anyway to handle it after an initial find_matches for versions specifically).

:thinking: I think, taking them into account would be helpful. Although, I'm not sure what you mean by the bugfix prevention here.

> We could consider signature validity + versions in the dependency resolver (a feature?), but backtracking if signature validation fails seems like it could be problematic (a change in behavior at least) - if someone prefers to install collections from AH but fall back to Galaxy for collections missing from AH; then any collections on AH that fail signature validation may not give an indicator something was wrong if an unsigned version of the collection was found on Galaxy.

Ah.. right. This would likely be considered a feature. Though, I wouldn't expect it to produce a different result, only to abort early.
Though, if we take into account the problem you described, it'd probably require some CLI/config toggles for the behavior and logging.
Still, I'd expect that if one requests signature verification, they'd want the entire dependency tree to be verified. Similar to pip's `--require-hashes` — if it's set for at least one distribution, it makes hashes mandatory for all transitive dependencies of the install session. This is what gave me the idea that the signature checking needs to be baked into the depresolver.

Just think about it:
- Collection A (signed on both AH and Galaxy) depends on a collection B (signed on AH but not signed on Galaxy)
- The latest versions are B==1.0.0 on AH and B==2.0.0 on Galaxy
- B==1.0.0 on AH depends on C==2.0.0 (signed on AH, missing from Galaxy)
- B==2.0.0 on Galaxy depends on D (only published on Galaxy and is compromised); D depends on an older version of C==1.0.0 that is present and signed on AH but has a security vulnerability

Now, when the signatures are not taken into account during the resolution, when installing A, it would go like this:
A -> B==2.0.0 (from Galaxy) -> D -> C==1.0.0 (this opens up possibilities for rollback attacks by attacking the supply chain first)

If the signatures are taken into account, then B==2.0.0 on Galaxy would be rejected and produce a different dependency tree which would arguably be more secure:
A -> B==1.0.0 (from AH) -> C==2.0.0 (a version of C with a vulnerability fixed)

I did some googling and even asked ChatGPT (it agrees with me!). Here's an interesting paper from TUF: https://theupdateframework.io/papers/package-management-security-tr08-02.pdf?raw=true — check out "3.2.5 Depends on Everything" which calls out the signature check on download, for example, or "3.2.6 Unsatisfiable Dependencies". Another one https://theupdateframework.io/papers/attacks-on-package-managers-ccs2008.pdf?raw=true also mentions these attack vectors in the "4.2 Package Signatures" section.

To summarize, now that I've done more research, it seems even more important to get it right...

Reply:
I've also asked a security engineering friend who confirmed that it's a good idea to take signatures into account to help prevent security flaws.

Reply:
It's good food for thought for the future at least!

I think I understand what you meant in your initial comment now though, so I can answer this more decisively:

> If the signatures are taken into consideration by the depresolver itself, wouldn't it cut the undesired backtracking path early?

No, only if there's a compromised collection, which should be an anomaly. This is fixing the general use case where everything is successfully installed (remember installing from Galaxy == no server signatures, but we query just in case because they might be added). We're making an extra Galaxy API call for every possible match, including those which might be backtracked later or won't be the 'latest'. Consider installing a single collection with no version specified. None of the versions are backtracked, but we fetch Galaxy signatures for every single match, not just the one we install.

Just to be clear, this was my implementation suggestion to avoid reintroducing the issue I'm fixing if you decide to add something like `--require-hashes` as an implicit feature:

* (perhaps) add an alternative to `resolve` that resolves + just returns latest 
* resolve the requirements normally (by version)
* validate the latest matches using signature
* If latest matches pass signature validation, return them
* Otherwise, mark the Resolution state with these new incompatibilities, and go back to step 2 using the new state

I'd be all for it without ^ caveat if you wanted to add it behind a toggle though.

To touch on download:

Signature verification happens when the artifact is extracted currently. This occurs for downloaded collections that are passed to `install`, so at the moment the burden is on the user to include any signatures from Galaxy as supplemental signatures. This was discussed a bit, just out of scope for the initial feature. This could perhaps be improved by inspecting tarfile contents + verifying before adding them to the download destination, or by saving signatures somewhere and including the files as supplemental signatures in the requirements.yml that's populated.

> I did some googling and even asked ChatGPT (it agrees with me!).

Haha, love it! I just wish it was better at being correct than sounding correct :smile: It totally gaslit me the last time I used it for something technical, it had me searching for a CLI option that never existed...

Reply:
Alright. Let's table the security hardening bit until we can properly discuss having it under a feature toggle...

With that in mind, I think that other restructuring suggestions sound reasonable, right? https://github.com/ansible/ansible/pull/80334/files#r1150826206 / https://github.com/ansible/ansible/pull/80334/files#r1150820606